README.md
# The voyage

[![CMake on a single platform](https://github.com/0x41337/voyage/actions/workflows/cmake-single-platform.yml/badge.svg)](https://github.com/0x41337/voyage/actions/workflows/cmake-single-platform.yml) [![CodeQL](https://github.com/0x41337/voyage/actions/workflows/codeql.yml/badge.svg)](https://github.com/0x41337/voyage/actions/workflows/codeql.yml) ![](https://img.shields.io/badge/License-GPL\_3.0-blue) ![](https://img.shields.io/badge/Version-0.4-yellow)

Voyage is a library for creating/managing hooks for unix systems.\
Voyage is inspired by the [Dobby](https://github.com/jmpews/Dobby), [Minhook](https://github.com/TsudaKageyu/minhook) and [Libmem](https://github.com/rdbo/libmem) libraries, which are great libraries. (You should look)\
If you know how to use some of the libraries mentioned, you will certainly feel comfortable using Voyage.

## Dependencies

* `googletest` Framework for writing tests.

## How to build

Make sure you have all dependencies that contain _\*_ in their name. You can build the library using cmake:

```sh
$ cmake -S . -B build
```

And

```sh
$ cmake --build build
```

If everything goes well you should have the shared library.

## Integration

Voyage is a C++ project so if you are creating a project with C/C++ together with CMake you can build Voyage alongside your project:

* Add Voyage as a vendorized dependency using git modules:

```sh
$ git submodule add repository/voyage.git vendor/voyage
```

> Note: You should change `repository` to the original repository. I just used this name to shorten the command and make it easier to read.

* Then add it to your CMake file:

```cmake
add_subdirectory(vendor/voyage)
```

* And lastly you can link your binary/library with Voyage:

```cmake
target_link_libraries(your_project voyage)
```

## The tests

Tests are written using the `googletest` testing framework and are located in the `tests` folder.

After compiling the project, you can use the command below to run the tests:

```sh
$ ctest --test-dir build/tests
```

## Development status

Voyage is still in the Beta phase in version 0.4. Below is a list of what has been implemented and what has not yet been implemented:

**Voyage core**

* [x] Trampoline
* [x] Memory protection
* [ ] Support for multiple architectures (Current focus)

**Voyage api**

* [x] `VE_FindSymbolAddress`
* [x] `VE_CreateHook`
* [x] `VE_RemoveHook`
* [x] `VE_EnableHook`
* [x] `VE_DisableHook`

**Voyage repository**

* [x] Documentation 
* [ ] Examples (Current focus)

## Documentation

All library documentation is available in [here](https://gabriels-organization-19.gitbook.io/voyage/)

## Examples

The examples are in the `examples` folder

## License and Terms

The Voyage code is under the `GPL-3.0` license. Third-party libraries used in the construction of the project may be under other licenses and terms, check the copy of the license and see the license and terms of the libraries used before using.


CHANGELOG
**Oct.04.2023**, Version `0.4` (_Beta_), b28be55b7893a7ce824c63ca2a941ed564abacd5

**Changes**:
- Refactoring multiple parts of the code. (Gabriel)
- Function creation: `VE_EnableHook`. (Gabriel)
- Function creation: `VE_Disablehook`. (Gabriel)
- Tests for `VE_EnableHook`. (Gabriel)
- Tests for `VE_Disablehook`. (Gabriel)


**Set.25.2023**, Version `0.3` (_Development_), 4a51877e0946bb0215162949690b8734b7678bd8

**Changes**:
- Fix "Stack smashing at `memcpy(&jmp[1], &offset, sizeof(offset));`" a bug that occurred because the JMP array was smaller than the instruction, causing a buffer overflow. (Gabriel)
- Function creation: `IsJumpToHook`. (Gabriel)
- Function creation: `VE_RemoveHook`. (Gabriel)
- Checking the target address to identify hooks in `VE_CreateHook`. (Gabriel)
- Added status return for: `Apply_detour()`. (Gabriel)
- Tests for `VE_RemoveHook`. (Gabriel)

**Set.24.2023**, Version `0.2` (_Development_), 60460f077b001dd5423280204170160bb07d738c

**Changes**:
- Capstone library removal. (Gabriel)
- Refactoring multiple parts of the code. (Gabriel)
- Function creation: `VE_CreateHook`. (Gabriel)
- Creation of `trampoline.hpp` and all code involving trampoline. (Gabriel)
- Tests for `VE_CreateHook`. (Gabriel)
- Basic example: `simple_hook.cpp`. (Gabriel)


tests/voyage_test.cpp
/*
  Voyage A library for creating and managing hooks for unix systems authored by Gabriel.
  <https://github.com/0x41337/voyage>     <041337@proton.me>

  Voyage is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Voyage is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <voyage.h>
#include <gtest/gtest.h>

/* Tests for: VE_FindSymbolAddress */
TEST(voyage_test, FindSymbolAddress_ModuleNotFound)
{
    void (*printf_ptr)(const char *_Format, ...) = nullptr;
    VE_STATUS status = VE_FindSymbolAddress("not_found", "printf", &printf_ptr);
    EXPECT_EQ(status, VE_ERROR_MODULE_NOT_FOUND);
};

TEST(voyage_test, FindSymbolAddress_SymbolNotFound)
{
    void (*printf_ptr)(const char *_Format, ...) = nullptr;
    VE_STATUS status = VE_FindSymbolAddress("", "not_found", &printf_ptr);
    EXPECT_EQ(status, VE_ERROR_SYMBOL_NOT_FOUND);
};

TEST(voyage_test, FindSymbolAddress_Success)
{
    void *(*printf_ptr)(const char *_Format, ...) = nullptr;
    VE_STATUS status = VE_FindSymbolAddress("", "printf", &printf_ptr);
    EXPECT_EQ(status, VE_OK);
    EXPECT_NE(printf_ptr, nullptr);
}

/* Tests for: VE_CreateHook */
int add(int x, int y) { return x + y; }
int add_hook(int x, int y) { return x - y; }
TEST(voyage_test, VE_CreateHook_Success)
{
    Hook hook;
    VE_STATUS status = VE_CreateHook(reinterpret_cast<void *>(&add), reinterpret_cast<void *>(&add_hook), &hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(add(1, 1), 0);
}

TEST(voyage_test, VE_CreateHook_HookAlreadyCreated)
{
    Hook hook;

    VE_CreateHook(reinterpret_cast<void *>(&add), reinterpret_cast<void *>(&add_hook), &hook);
    VE_STATUS status = VE_CreateHook(reinterpret_cast<void *>(&add), reinterpret_cast<void *>(&add_hook), &hook);

    EXPECT_EQ(status, VE_ERROR_ALREADY_CREATED);
}

/* tests for: VE_RemoveHook */
int sub(int x, int y) { return x - y; };
int sub_hook(int x, int y) { return x + y; };
TEST(voyage_test, VE_RemoveHook_Success)
{
    Hook hook;

    VE_STATUS status = VE_CreateHook(reinterpret_cast<void *>(&sub), reinterpret_cast<void *>(&sub_hook), &hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 2);

    status = VE_RemoveHook(&hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 0);
}

/* tests for: VE_EnableHook and VE_DisableHook */
TEST(voyage_test, VE_DisableHook_Success)
{
    Hook hook;

    VE_STATUS status = VE_CreateHook(reinterpret_cast<void *>(&sub), reinterpret_cast<void *>(&sub_hook), &hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 2);

    status = VE_DisableHook(&hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 0);
}

TEST(voyage_test, VE_EnableHook_Success)
{
    Hook hook;

    VE_STATUS status = VE_CreateHook(reinterpret_cast<void *>(&sub), reinterpret_cast<void *>(&sub_hook), &hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 2);

    status = VE_DisableHook(&hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 0);

    status = VE_EnableHook(&hook);
    EXPECT_EQ(status, VE_OK);
    EXPECT_EQ(sub(1, 1), 2);
}

src/voyage.cpp
/*
  Voyage A library for creating and managing hooks for unix systems authored by Gabriel.
  <https://github.com/0x41337/voyage>     <041337@proton.me>

  Voyage is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Voyage is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <voyage.h>
#include <trampoline.hpp>

#include <dlfcn.h>
#include <memory.h>

/// @brief Search by symbol address
/// @param img_name Image/module name
/// @param sym_name Symbol name
/// @param dest Destination that will receive the found address
/// @return `VE_STATUS` can be [`VE_ERROR_MODULE_NOT_FOUND`, `VE_ERROR_SYMBOL_NOT_FOUND` and `VE_OK`]
VE_STATUS VE_FindSymbolAddress(const char *img_name, const char *sym_name, void *dest)
{
    // Find the image
    void *handle = dlopen(img_name, RTLD_NOW | RTLD_GLOBAL);
    if (handle == nullptr)
    {
        return VE_ERROR_MODULE_NOT_FOUND;
    }

    // Find symbol in the image
    void *symbol = dlsym(handle, sym_name);
    if (symbol == nullptr)
    {
        return VE_ERROR_SYMBOL_NOT_FOUND;
    }

    // Copy address of symbol to destination
    memcpy(dest, &symbol, sizeof(void *));

    // Close module
    dlclose(handle);

    return VE_OK;
}

/// @brief Create a hook
/// @param target_address Address to the target
/// @param hook_address Address to the hook
/// @return `VE_STATUS` can be [`VE_ERROR_ALREADY_CREATED`, `VE_ERROR_ALLOCATE_MEMORY`, `VE_ERROR_NOT_CREATED`, `VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS VE_CreateHook(void *target_address, void *hook_address, Hook *dest)
{
    // Check if there is already a hook for this address
    if (IsJumpToHook(target_address, hook_address))
        return VE_ERROR_ALREADY_CREATED;

    // Ensures that the memory page is writable and executable
    auto status = Remove_memory_protection(target_address);
    if (status != VE_OK)
        return status;

    // Save the original instructions
    Save_original_instructions(target_address, dest);

    // Apply JMP injection to the target address
    status = Apply_detour(target_address, hook_address);
    if (status != VE_OK)
        return status;

    // Restores memory protection
    status = Restore_memory_protection(target_address);
    if (status != VE_OK)
        return status;

    // Configure the hook
    dest->target = target_address;
    dest->hook_address = hook_address;
    dest->enabled = true;

    return VE_OK;
}

/// @brief Remove a hook completely
/// @param hook The strutct that represents the hook
/// @return `VE_STATUS` can be [VE_ERROR_MEMORY_PROTECT and VE_OK]
VE_STATUS VE_RemoveHook(Hook *hook)
{
    // Ensures that the memory page is writable and executable
    auto status = Remove_memory_protection(hook->target);
    if (status != VE_OK)
        return status;

    // Restore original instructions
    Restore_original_instructions(hook);

    // Restores memory protection
    status = Restore_memory_protection(hook->target);
    if (status != VE_OK)
        return status;

    return VE_OK;
}

/// @brief Enable a hook
/// @param hook The hook
/// @return `VE_STATUS` can be [`VE_ERROR_ALREADY_ENABLED`, `VE_ERROR_MEMORY_PROTECT`, `VE_ERROR_ALLOCATE_MEMORY` and `VE_OK`]
VE_STATUS VE_EnableHook(Hook *hook)
{
    // Check if hook already enabled
    if (hook->enabled)
        return VE_ERROR_ALREADY_ENABLED;

    // Ensures that the memory page is writable and executable
    auto status = Remove_memory_protection(hook->target);
    if (status != VE_OK)
        return status;

    // Apply JMP injection to the target address
    status = Apply_detour(hook->target, hook->hook_address);
    if (status != VE_OK)
        return status;

    // Restores memory protection
    status = Restore_memory_protection(hook->target);
    if (status != VE_OK)
        return status;

    // Configure the hook
    hook->enabled = true;

    return VE_OK;
}

/// @brief Disable a hook
/// @param hook The hook
/// @return `VE_STATUS` can be [`VE_ERROR_ALREADY_DISABLED`, `VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS VE_DisableHook(Hook *hook)
{
    // Check if hook already enabled
    if (!hook->enabled)
        return VE_ERROR_ALREADY_DISABLED;

    // Ensures that the memory page is writable and executable
    auto status = Remove_memory_protection(hook->target);
    if (status != VE_OK)
        return status;

    // Restore original instructions
    Restore_original_instructions(hook);

    // Restores memory protection
    status = Restore_memory_protection(hook->target);
    if (status != VE_OK)
        return status;

    // Configure the hook
    hook->enabled = false;

    return VE_OK;
}


src/trampoline.hpp
/*
  Voyage A library for creating and managing hooks for unix systems authored by Gabriel.
  <https://github.com/0x41337/voyage>     <041337@proton.me>

  Voyage is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Voyage is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#pragma once

#include <unistd.h>
#include <memory.h>
#include <malloc.h>
#include <inttypes.h>
#include <sys/mman.h>

#include <voyage.h>

/// @brief Removes target address memory protection
/// @param target_address  Address to the target
/// @return VE_STATUS can be [`VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS Remove_memory_protection(void *target_address)
{
    if (mprotect(reinterpret_cast<void *>((uintptr_t)target_address & ~(sysconf(_SC_PAGE_SIZE) - 1)), sysconf(_SC_PAGE_SIZE), PROT_READ | PROT_WRITE | PROT_EXEC) == -1)
        return VE_ERROR_MEMORY_PROTECT;

    return VE_OK;
}

/// @brief Restores target address memory protection
/// @param target_address Address to the target
/// @return VE_STATUS can be [`VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS Restore_memory_protection(void *target_address)
{
    if (mprotect(reinterpret_cast<void *>((uintptr_t)target_address & ~(sysconf(_SC_PAGE_SIZE) - 1)), sysconf(_SC_PAGE_SIZE), PROT_READ | PROT_EXEC) == -1)
        return VE_ERROR_MEMORY_PROTECT;

    return VE_OK;
}

/// @brief Saves the original instructions that are at the target address
/// @param target_address Address to the target
/// @param dest Pointer to hook
void Save_original_instructions(void *target_address, Hook *dest)
{
    memcpy(dest->original_instructions, target_address, sizeof(dest->original_instructions));
}

/// @brief Restores instructions that were saved with Save_original_instructions
/// @param hook The hook
void Restore_original_instructions(Hook *hook)
{
    memcpy(hook->target, hook->original_instructions, sizeof(hook->original_instructions));
}

/// @brief Reads instructions from the target address and checks if it is a `jmp`
/// @param target_address Address to the target
/// @param hook_address Address to the hook
/// @return If the address contains a jmp instruction
bool IsJumpToHook(void *target_address, void *hook_address)
{
    // Configure the jump size and buffer
    const int jmp_size = 5;
    uint8_t buffer[jmp_size];

    // Read the first bytes of memory at target_address
    memcpy(buffer, target_address, jmp_size);

    // Checks if the first bytes match the jmp instruction
    if (buffer[0] == 0xE9)
    {
        // Calculates the hop destination address
        int32_t jump_offset;
        memcpy(&jump_offset, buffer + 1, sizeof(jump_offset));

        // Calculates absolute destination address
        void *jump_target = reinterpret_cast<void *>(reinterpret_cast<uintptr_t>(target_address) + 5 + jump_offset);

        // Checks if the destination address matches the hook_address
        return jump_target == hook_address;
    }

    return false;
}

/// @brief Injects the `jmp` instruction at the `target address` to the `hook address`
/// @param target_address Address to the target
/// @param hook_address Address to the hook
/// @return VE_STATUS can be [`VE_ERROR_ALLOCATE_MEMORY` and `VE_OK`]
VE_STATUS Apply_detour(void *target_address, void *hook_address)
{
    // Calculate offset
    intptr_t offset = ((intptr_t)hook_address - (intptr_t)target_address - 5);

    // Allocate memory for the jump instruction
    unsigned char *jmp = reinterpret_cast<unsigned char *>(malloc(5));
    if (jmp == NULL)
        return VE_ERROR_ALLOCATE_MEMORY;

    // Copy the jump instruction to the allocated memory
    memcpy(jmp, "\xE9", 1);
    memcpy(jmp + 1, &offset, sizeof(offset));

    // Apply the detour
    memcpy(target_address, jmp, 5);

    // Free the allocated memory
    free(jmp);

    return VE_OK;
}


include/voyage.h
/*
  Voyage A library for creating and managing hooks for unix systems authored by Gabriel.
  <https://github.com/0x41337/voyage>     <041337@proton.me>

  Voyage is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Voyage is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#pragma once

/// @brief Represents a hook
struct Hook
{
  void *target;
  void *hook_address;
  unsigned char original_instructions[5];
  bool enabled;
};

/// @brief All API status codes
typedef enum VE_STATUS
{
  VE_OK = 0,

  // Creating hooks errors
  VE_ERROR_ALREADY_CREATED,
  VE_ERROR_NOT_CREATED,

  // Enable hooks errors
  VE_ERROR_ALREADY_ENABLED,
  VE_ERROR_ALREADY_DISABLED,

  // Memory protection errors
  VE_ERROR_MEMORY_PROTECT,

  // Memory allocation errors
  VE_ERROR_ALLOCATE_MEMORY,

  // Find symbol erros
  VE_ERROR_MODULE_NOT_FOUND,
  VE_ERROR_SYMBOL_NOT_FOUND,

} VE_STATUS;

// Constants
#define VE_VERSION 0.4;

/// @brief Search by symbol address
/// @param img_name Image/module name
/// @param sym_name Symbol name
/// @param dest Destination that will receive the found address
/// @return `VE_STATUS` can be [`VE_ERROR_MODULE_NOT_FOUND`, `VE_ERROR_SYMBOL_NOT_FOUND` and `VE_OK`]
VE_STATUS VE_FindSymbolAddress(const char *img_name, const char *sym_name, void *dest);

/// @brief Create a hook
/// @param target_address Address to the target
/// @param hook_address Address to the hook
/// @param original_function  Pointer to original function
/// @return `VE_STATUS` can be [`VE_ERROR_ALREADY_CREATED`, `VE_ERROR_NOT_CREATED`, `VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS VE_CreateHook(void *target_address, void *hook_address, Hook *dest);

/// @brief Remove a hook completely
/// @param hook The strutct that represents the hook
/// @return `VE_STATUS` can be [`VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS VE_RemoveHook(Hook *hook);

/// @brief Enable a hook
/// @param hook The hook
/// @return `VE_STATUS` can be [`VE_ERROR_ALREADY_ENABLED`, `VE_ERROR_MEMORY_PROTECT`, `VE_ERROR_ALLOCATE_MEMORY` and `VE_OK`]
VE_STATUS VE_EnableHook(Hook *hook);

/// @brief Disable a hook
/// @param hook The hook
/// @return `VE_STATUS` can be [`VE_ERROR_ALREADY_DISABLED`, `VE_ERROR_MEMORY_PROTECT` and `VE_OK`]
VE_STATUS VE_DisableHook(Hook *hook);
